java/json/src/main/java/org/bondlib/SimpleJsonReader.java

```diff
@@ -0,0 +1,480 @@
+    }
+
+    @Override
+    public final SimpleJsonReader cloneProtocolReader() throws IOException {
+        // no support for cloning (which is required for Bonded functionality)
+        throw new UnsupportedOperationException("Cloning is not implemented.");
+    }
+
+    /**
+     * Advances the JSON parser to the next token and checks whether the end of input has been reached
+     * (in which case the method throws an exception). Returns a non-null token.
+     */
+    private JsonToken advanceParserToNextToken() throws IOException {
+        final JsonToken currentToken;
+        if (this.hasPeekedParserForNextToken) {
+            // if has already peeked the next token (i.e. peekParserForNextToken was called),
+            // then consume the current token without advancing the parser, which will
+            // be advanced the next time this method (or peekParserForNextToken) is called
+            currentToken = this.parser.currentToken();
+            this.hasPeekedParserForNextToken = false;
+        } else {
+            // otherwise, advance to the next token
+            try {
+                currentToken = this.parser.nextToken();
+            } catch (JsonProcessingException e) {
+                // wrap Jackson exception by a Bond exception
+                throw createInvalidBondDataException(e, "Error processing JSON input: " + e.getMessage());
+            }
+        }
+        // check the current token for the end of stream
+        if (currentToken == null) {
+            throw new EOFException("The end of JSON input has been reached.");
+        }
+        return currentToken;
+    }
+
+    /**
+     * Peeks the next token within the JSON parser and returns it (possibly null).
+     * Can peek a token only once; subsequent calls will return the same token
+     * until {@link #advanceParserToNextToken} is called to advance the parser.
+     */
+    private JsonToken peekParserForNextToken() throws IOException {
+        final JsonToken nextToken;
+        if (this.hasPeekedParserForNextToken) {
+            // if has already peeked once, consume the current token
+            nextToken = this.parser.currentToken();
+        } else {
+            // otherwise, advance the parser and get the next token
+            // which may be null (when the end of the input is reached)
+            try {
+                nextToken = this.parser.nextToken();
+            } catch (JsonProcessingException e) {
+                // wrap Jackson exception by a Bond exception
+                throw createInvalidBondDataException(e, "Error processing JSON input: " + e.getMessage());
+            }
+            this.hasPeekedParserForNextToken = true;
+        }
+        return nextToken;
+    }
+
+    /**
+     * Creates a new {@link InvalidBondDataException} exception which indicates a Bond-specific protocol error.
+     */
+    private static InvalidBondDataException createInvalidBondDataException(
```

Comment:
If `core` has a `Throw`, maybe `json` should, too?

Reply:
I now think actually throwing in a helper is not that elegant, and we just create exception but throw elsewhere. This way you don't need to deal with bogus return values, etc. 

I won't change Throw itself for now.

---
In reply to: [143861904](https://github.com/Microsoft/bond/pull/632#discussion_r143861904) [](ancestors = 143861904)